chapter3/challenges/program/lib/password_manager_2.rb

```diff
@@ -1,38 +1,64 @@
+require 'date'
 class PasswordManager2
   def initialize
+    @manager = []
+    @used_passwords = []
+    @used_services = []
   end
 
   def add(service, password)
     # Takes two strings, the service and its password, as arguments
     # Stores the new service and its password
     # Unless the new password is already in use elsewhere
     # Or the service already has a stored password
+    if @used_passwords.include?(password)
+      return "ERROR: Passwords must be unique"
+    elsif @used_services.include?(service)
+      return "ERROR: Service names must be unique"
+    else
+      @manager << { "service" => service, "password" => password, "added_on" => Date.today }
```

Comment:
You  _could_ use DateTime.now() if looking for more precision but otherwise nice. Also maybe formatting, you could do:

`@manager << { 
"service" => service, 
"password" => password, 
"added_on" => Date.today 
}`

Just for readability?

Reply:
That hasn't formatted correctly!

Reply:
I meant putting each key/value on a new line

Reply:
I like the way you did it in your version by declaring a new variable called new_entry